source/common/http/conn_manager_impl.cc

```diff
@@ -509,13 +509,25 @@ void ConnectionManagerImpl::chargeTracingStats(const Tracing::Reason& tracing_re
   }
 }
 
-void ConnectionManagerImpl::RdsRouteConfigUpdateRequester::requestRouteConfigUpdate(
+void ConnectionManagerImpl::RdsRouteConfigUpdateRequester::requestVhdsUpdate(
     const std::string host_header, Event::Dispatcher& thread_local_dispatcher,
     Http::RouteConfigUpdatedCallbackSharedPtr route_config_updated_cb) {
   route_config_provider_->requestVirtualHostsUpdate(host_header, thread_local_dispatcher,
                                                     std::move(route_config_updated_cb));
 }
 
+void ConnectionManagerImpl::RdsRouteConfigUpdateRequester::requestSrdsUpdate(
+    uint64_t key_hash, Event::Dispatcher& thread_local_dispatcher,
+    Http::RouteConfigUpdatedCallback route_config_updated_cb) {
+  // If it is inline scope_route_config_provider, will be cast to nullptr.
+  if (!scoped_route_config_provider_) {
+    thread_local_dispatcher.post([route_config_updated_cb] { route_config_updated_cb(false); });
```

Comment:
It has to be. Because if we execute it directly in the current call stack, the state of the active_stream will be wrong. We post the callback so that it can be executed after exiting the current call stack and finishing the on demand filter.

Reply:
Can you elaborate? In general, Envoy style really discourages posting callbacks to the same thread and this continuation style passing, since it creates extremely hard to reason about logic. Is there a way to restructure this?

Reply:
Sure, I can do it in a different style.